### PR TITLE
Fixing ngram masking in training data preprocessing

### PR DESCRIPTION
**Pull request by thomwolf**
Hi, thanks for the great work and for taking the time to document the code and training/fine-tuning scripts and commands. You guys rocks!

While converting the code for [pytorch-pretrained-bert](https://github.com/huggingface/pytorch-pretrained-BERT), I noticed a small bug in the data pre-processing logic of `data_utils.py` which made the script masks `n` consecutive sub-words instead of the intended behavior of masking `n` consecutive full words (n-grams).

This PR fix this.

### CODE REVIEW

**Comment by kimiyoung** [comment]
@thomwolf Thanks for spotting this out! It is indeed not what we intended, but seems to work fine. We are quickly testing the performance of using true ngram masking and will get back shortly.

**Comment by thomwolf** [comment]
Some thoughts on this if you don't mind me talking about science/tech in the PR (not the best place I guess):

If I get it right, it might probably not change much because the `g` state of a masked sub-word token inside an n-gram will still have access to the `h` state of previous sub-word tokens in the same masked n-gram so this is not really a "full-word" or "full n-gram" masking and you may not get the benefits shown in BERT's [Whole word masking variant](https://github.com/google-research/bert) for instance. To have full word masking, the `non_tgt_mask` should be tweaked to mask all the tokens in the consecutive n-grams which is probably a bit cumbersome to do.

On another note, we've been playing a bit with two-stream attention concurrently to your work (not enough to scale the training though) and the way we did it was, instead of doing successive attention, was to add the target tokens at the end and design a custom attention mask like this (this way you can keep the exact same architecture). It might be more difficult to do with relative positioning than absolute positions, though. Here is an example doing full n-gram masking with `h` and `g` in a parallel setup if you're interested:

masking_xlnet

Ok, end of my 2 cents :-)

**Comment by kimiyoung** [comment]
Thanks for sharing your thoughts. I have a different viewpoint. There are two types of dependencies to model, intra-word dependency and inter-word dependency.

For BERT, using whole word masking emphasizes inter-word dependency. For XLNet, inter-word dependency is captured by p(first token of word | other words) and in part by p(other tokens of word | other words, prior tokens in permutation), while the latter also captures intra-word dependency. I personally like the elegance of the auto-regressive framework because it naturally captures all dependencies you want. That said, adding ngram (subword) masking still improves XLNet because it enables modeling both dependencies.

At the end of the day, we can only count on experimental results given the lack of theories, so I think it will still be worth it to experiment with removing the intra-word dependency. That could be achieved by modifying the `_local_perm` function with an additional input indicating word spans.

**Comment by SivilTaram** [comment]
@kimiyoung Thanks for the great work! I'm fond of and curious about design principle behind the two-steam attention. From my personal point of view, @thomwolf argues about the `full mask` applied in BERT, which prevents the information from masked words/sub-words. In contrast, the attention `g` in XLNet actually observes the leaked information of masked words from previous `h`. If I get it right, the information flow is just like `h(1)->h(2)->g(1)`. I do not know if it is designed for intention?

Besides it, the explanation of intra and inter word dependency from your point is also attractive! Looking forward to your updated experimental results!

**Comment by thomwolf** [comment]
Thanks for the feedback @kimiyoung. I see your point with regards to losing the auto-regressive formulation when we mask consecutive sub-tokens. The distinction of intra-word dependency and inter-word dependency is a nice way to formulate this indeed.

**Comment by kimiyoung** [comment]
@thomwolf Thanks for the code! It seems to me that there is some [pytorch code](https://github.com/zihangdai/xlnet/pull/16/files#diff-bbb73959f56c6913089a6a4d99430eafR737) in your commits. Is it possible to clean those a bit?

**Comment by thomwolf** [comment]
Oh sorry, I kept playing with this fork and I forgot the PR was not merged. I'll do another PR.
